R/tbl_split.R

```diff
@@ -40,27 +129,190 @@ tbl_split.gtsummary <- function(x, variables, ...) {
+        structure(class = c("tbl_split", "list"))
+    )
+  }
+
+  # check inputs ---------------------------------------------------------------
+  check_class(x, "gtsummary")
+  keys <- maybe_missing(keys, default = first_column(x))
+  cards::process_selectors(x$table_body, keys = {{ keys }})
+  check_class(groups, "list")
+  cards::check_list_elements(
+    groups,
+    predicate = is.character,
+    error_msg = "Each element of the {.arg groups} argument's list must be a {.cls character} vector."
+  )
+
+  # check all variables in groups appear in the table
+  columns_do_not_exist <-
+    unlist(groups) |>
+    setdiff(x$table_styling$header$column)
+  if (!is_empty(columns_do_not_exist)) {
+    cli::cli_abort()
+  }
+
+  # check that every un-hidden column is present
+  missing_cols <-
+    x$table_styling$header[!x$table_styling$header$hide, ][["column"]] |>
+    setdiff(c(keys, unlist(groups)))
+  if (!is_empty(missing_cols)) {
+    cli::cli_inform(
+      c("The following columns were not listed in either {.arg keys} or {.arg groups} argument: {.val {missing_cols}}",
+        "i" = "These columns have been added to the end of {.arg groups}.",
+        "*" = "Run {.fun gtsummary::show_header_names} for a list of all column names."
+      )
+    )
+    groups <- c(groups, list(missing_cols))
+  }
+
+  # footnotes and caption check
+  footnotes <- rlang::arg_match(footnotes)
+  caption <- rlang::arg_match(caption)
+
+  # splitting table ------------------------------------------------------------
+  tbl_list <- vector(mode = "list", length = length(groups))
+  for (i in seq_along(groups)) {
+    tbl_list[[i]] <- gtsummary::modify_column_hide(x, columns = -all_of(union(keys, groups[[i]])))
+  }
+
+  # caption/footnotes handling -----------------------------------------------
+  if (length(tbl_list) > 1) {
+    tbl_list <- .modify_footnotes_caption(tbl_list, footnotes, caption)
+  }
+
+  # return list of tbls --------------------------------------------------------
+  tbl_list |>
+    structure(class = c("tbl_split", "list"))
+}
+
+# helper function for handling caption/footnotes
+.modify_footnotes_caption <- function(tbl_list, footnotes, caption) {
+  which_to_modify <- if (footnotes == "first") {
+    seq(2, length(tbl_list))
+  } else if (footnotes == "last") {
+    seq(1, length(tbl_list) - 1)
+  } else {
+    NULL
+  }
+  if (!is.null(which_to_modify)) {
+    for (i in which_to_modify) {
+      tbl_list[[i]] <- tbl_list[[i]] |>
+        remove_source_note() |>
+        remove_abbreviation()
+    }
+  }
+  if (caption == "first") {
+    # Changing the call_list does not affect the output, should it be trimmed?
+    for (i in seq(2, length(tbl_list))) {
+      tbl_list[[i]]$table_styling$caption <- NULL
+    }
+  }
+  tbl_list
+}
+
+#' @export
+#' @rdname tbl_split_by
+first_column <- function(x) {
+  set_cli_abort_call()
+  check_class(x, "gtsummary")
+
+  x$table_styling$header[!x$table_styling$header$hide, ][["column"]][1]
 }
```

Comment:
I don't think we need to export this function. We have an internal function in gtsummary called `.first_unhidden_column()` with the same functionality, and the user never sees the calls to the function.

Reply:
done!